File: ETTrace/ETTraceRunner/JSONWrapper.m
//
//  TestClass.m
//  ETTraceRunner
//
//  Created by Noah Martin on 4/13/23.
//

#import <Foundation/Foundation.h>
#import "JSONWrapper.h"

@implementation JSONWrapper

+ (NSDictionary *)toDictionary:(FlameNode *)node {
  NSObject *children;
  if (node.children.count == 1) {
    children = [JSONWrapper toDictionary:node.children[0]];
  } else {
    children = [[NSMutableArray alloc] init];
    for (FlameNode * c in node.children) {
      [(NSMutableArray *) children addObject:[JSONWrapper toDictionary:c]];
    }
  }

  return @{
    @"name": node.name,
    @"start": @(node.start),
    @"duration": @(node.duration),
    @"library": node.library ? node.library : @"",
    @"children": children,
  };
}

+ (NSData *)toData:(FlameNode *)input {
  return [NSJSONSerialization dataWithJSONObject:[JSONWrapper toDictionary:input] options:NSJSONWritingWithoutEscapingSlashes error:nil];
}

@end

File: ETTrace/ETTraceRunner/JSONWrapper.swift
//
//  File.swift
//  
//
//  Created by Itay Brenner on 6/6/23.
//

import Foundation
import ETModels

class JSONWrapper {
    static func toDictionary(_ node: FlameNode) -> NSDictionary {
        var children: NSObject
        if node.children.count == 1 {
            children = toDictionary(node.children[0])
        } else {
            let childArray = NSMutableArray()
            for c in node.children {
                childArray.add(toDictionary(c))
            }
            children = childArray
        }
        
        return [
            "name": node.name,
            "start": node.start,
            "duration": node.duration,
            "library": node.library ?? "",
            "children": children
        ]
    }
    
    static func toData(_ input: FlameNode) -> Data? {
        let jsonObject = toDictionary(input)
        return try? JSONSerialization.data(withJSONObject: jsonObject, options: .withoutEscapingSlashes)
    }
}
